File: python3/min_cost_climbing_stairs.py
# leetcode

class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        leng = len(cost)
        dp = [0] * leng
        
        for idx in range(leng):
            if idx < 2:
                dp[idx] = cost[idx]
            else:
                dp[idx] = cost[idx] + min(dp[idx-1], dp[idx-2])
            idx += 1
                
        print(dp)
        return min(dp[leng-1], dp[leng-2])

File: golang/min_cost_climbing_stairs.go
package main

// Space : O(n)
// Time  : O(n)

func minCostClimbingStairs(cost []int) int {
    n := len(cost)
    var dp []int
    
    for i := 0; i < n; i++ {
        dp = append(dp, 0)
    }
    
    for i := 0; i < n; i++ {
        if i < 2 {
            dp[i] = cost[i]
        } else {
            dp[i] = cost[i] + Min(dp[i-2], dp[i-1])
        }
    }
    
    return Min(dp[n-2], dp[n-1])
}